src/ripple/app/tx/impl/Payment.cpp

```diff
@@ -270,27 +270,17 @@ Payment::preclaim(PreclaimContext const& ctx)
         return tecDST_TAG_NEEDED;
     }
 
-    if (paths || sendMax || !saDstAmount.native())
+    // Payment with at least one intermediate step and uses transitive balances.
+    if ((paths || sendMax || !saDstAmount.native()) && ctx.view.open())
     {
-        // Ripple payment with at least one intermediate step and uses
-        // transitive balances.
-
-        // Copy paths into an editable class.
-        STPathSet const spsPaths = ctx.tx.getFieldPathSet(sfPaths);
-
-        auto pathTooBig = spsPaths.size() > MaxPathSize;
+        STPathSet const& paths = ctx.tx.getFieldPathSet(sfPaths);
 
-        if (!pathTooBig)
-            for (auto const& path : spsPaths)
-                if (path.size() > MaxPathLength)
-                {
-                    pathTooBig = true;
-                    break;
-                }
-
-        if (ctx.view.open() && pathTooBig)
+        if (paths.size() > MaxPathSize ||
+            std::any_of(paths.begin(), paths.end(), [](STPath const& path) {
+                return path.size() > MaxPathSize;
```

Comment:
Bug: the check against individual paths should compare against `MaxPathLength` here.